File: frontend/src/types/habit.ts
export interface Habit {
  id: number;
  user_id: number;
  tracking_period_id: number;
  name: string;
  description: string | null;
  start_date: string;
  end_date: string | null;
  is_active: boolean;
  created_at: string;
}

export interface HabitCreate {
  tracking_period_id: number;
  name: string;
  description?: string;
  start_date: string;
  end_date?: string;
}

File: backend/app/schemas/habit.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class HabitCreate(BaseModel):
    tracking_period_id: int

    name: str = Field(
        min_length=1,
        max_length=100
    )

    description: str | None = Field(
        default=None,
        max_length=500
    )

    start_date: date

    end_date: date | None = None


class HabitResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    tracking_period_id: int
    name: str
    description: str | None
    start_date: date
    end_date: date | None
    is_active: bool
    created_at: datetime


class HabitUpdate(BaseModel):
    tracking_period_id: int

    name: str = Field(
        min_length=1,
        max_length=100
    )

    description: str | None = Field(
        default=None,
        max_length=500
    )

    start_date: date

    end_date: date | None = None

    is_active: bool